src/platform/datapath_raw_socket.c

```diff
@@ -335,52 +383,44 @@ CxPlatResolveRoute(
     // First call GetIpNetEntry2 to see if there's already a cached neighbor. If there
     // isn't one, or if the cached neighbor's state is unreachable (which, NB, can happen
     // in the case where a route lookup resulted in a dummy neighbor entry being created
-    // in TCPIP.sys) or incomplete, then do a solicitation.
+    // in TCPIP.sys) or incomplete, then queue up a solicitation event.
     //
     Status = GetIpNetEntry2(&IpnetRow);
+    QuicTraceLogConnInfo(
+        RouteResolutionStart,
+        Context,
+        "Starting to look up neighbor on Path[%hhu] with status %u",
+        PathId,
+        Status);
     if (Status != ERROR_SUCCESS || IpnetRow.State <= NlnsIncomplete) {
-        Status =
-            ResolveIpNetEntry2(
-                &IpnetRow,
-                &Route->LocalAddress);
-        if (Status != 0) {
+        CXPLAT_ROUTE_RESOLUTION_WORKER* Worker = Socket->Datapath->RouteResolutionWorker;
+        CXPLAT_ROUTE_RESOLUTION_OPERATION* Operation = CxPlatPoolAlloc(&Worker->OperationPool);
+        if (Operation == NULL) {
             QuicTraceEvent(
-                DatapathErrorStatus,
-                "[data][%p] ERROR, %u, %s.",
-                Socket,
-                Status,
-                "ResolveIpNetEntry2");
+                AllocFailure,
+                "Allocation of '%s' failed. (%llu bytes)",
+                "CXPLAT_DATAPATH",
+                sizeof(CXPLAT_ROUTE_RESOLUTION_OPERATION));
+            Status = ERROR_NOT_ENOUGH_MEMORY;
             goto Done;
         }
-    }
-    CXPLAT_DBG_ASSERT(IpnetRow.PhysicalAddressLength == sizeof(Route->NextHopLinkLayerAddress));
-    CxPlatCopyMemory(&Route->NextHopLinkLayerAddress, IpnetRow.PhysicalAddress, sizeof(Route->NextHopLinkLayerAddress));
-
-    //
-    // Find the interface that matches the route we just looked up.
-    //
-    CXPLAT_LIST_ENTRY* Entry = Socket->Datapath->Interfaces.Flink;
-    for (; Entry != &Socket->Datapath->Interfaces; Entry = Entry->Flink) {
-        CXPLAT_INTERFACE* Interface = CONTAINING_RECORD(Entry, CXPLAT_INTERFACE, Link);
-        if (Interface->IfIndex == IpforwardRow.InterfaceIndex) {
-            CxPlatDpRawAssignQueue(Interface, Route);
-            break;
-        }
-    }
-
-    if (Route->Queue == NULL) {
-        Status = QUIC_STATUS_NOT_FOUND;
-        QuicTraceEvent(
-            DatapathError,
-            "[data][%p] ERROR, %s.",
-            Socket,
-            "no matching interface/queue");
-        goto Done;
+        Operation->IpnetRow = IpnetRow;
+        Operation->Context = Context;
+        Operation->Callback = Callback;
+        Operation->PathId = PathId;
+        CxPlatDispatchLockAcquire(&Worker->Lock);
+        CxPlatListInsertTail(&Worker->Operations, &Operation->WorkerLink);
+        CxPlatDispatchLockRelease(&Worker->Lock);
+        CxPlatEventSet(Worker->Ready);
+        Status = ERROR_IO_PENDING;
```

Comment:
We have a `QUIC_STATUS_PENDING`.

Reply:
Yeah, but right before we return status, we convert it to hresult. `return HRESULT_FROM_WIN32(Status);`